File: warehouse_app/lib/models/attendance.dart
class Attendance {
  final int? id;
  final int userId;
  final String username;
  final String? fullName;
  final DateTime checkInTime;
  final DateTime? checkOutTime;
  final String status;
  final String? checkInPhotoUrl;
  final String? checkOutPhotoUrl;
  final String? checkInLocation;
  final String? checkOutLocation;
  final double? workingHours;
  final double? faceRecognitionConfidence;
  final bool isVerified;
  final String? notes;

  Attendance({
    this.id,
    required this.userId,
    required this.username,
    this.fullName,
    required this.checkInTime,
    this.checkOutTime,
    required this.status,
    this.checkInPhotoUrl,
    this.checkOutPhotoUrl,
    this.checkInLocation,
    this.checkOutLocation,
    this.workingHours,
    this.faceRecognitionConfidence,
    required this.isVerified,
    this.notes,
  });

  factory Attendance.fromJson(Map<String, dynamic> json) {
    return Attendance(
      id: json['id'],
      userId: json['userId'],
      username: json['username'],
      fullName: json['fullName'],
      checkInTime: DateTime.parse(json['checkInTime']),
      checkOutTime: json['checkOutTime'] != null 
          ? DateTime.parse(json['checkOutTime']) 
          : null,
      status: json['status'],
      checkInPhotoUrl: json['checkInPhotoUrl'],
      checkOutPhotoUrl: json['checkOutPhotoUrl'],
      checkInLocation: json['checkInLocation'],
      checkOutLocation: json['checkOutLocation'],
      workingHours: json['workingHours']?.toDouble(),
      faceRecognitionConfidence: json['faceRecognitionConfidence']?.toDouble(),
      isVerified: json['isVerified'] ?? true,
      notes: json['notes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'username': username,
      'fullName': fullName,
      'checkInTime': checkInTime.toIso8601String(),
      'checkOutTime': checkOutTime?.toIso8601String(),
      'status': status,
      'checkInPhotoUrl': checkInPhotoUrl,
      'checkOutPhotoUrl': checkOutPhotoUrl,
      'checkInLocation': checkInLocation,
      'checkOutLocation': checkOutLocation,
      'workingHours': workingHours,
      'faceRecognitionConfidence': faceRecognitionConfidence,
      'isVerified': isVerified,
      'notes': notes,
    };
  }

  bool get hasCheckedOut => checkOutTime != null;

  String get statusDisplay {
    switch (status) {
      case 'ON_TIME':
        return 'Đúng giờ';
      case 'LATE':
        return 'Muộn';
      case 'EARLY_LEAVE':
        return 'Về sớm';
      case 'ABSENT':
        return 'Vắng';
      default:
        return status;
    }
  }
}

File: warehouse-backend/src/main/java/com/warehousebackend/entity/Attendance.java
package com.warehousebackend.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "attendance")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Attendance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "check_in_time", nullable = false)
    private LocalDateTime checkInTime;

    @Column(name = "check_out_time")
    private LocalDateTime checkOutTime;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private AttendanceStatus status; // ON_TIME, LATE, EARLY_LEAVE, ABSENT

    @Column(name = "check_in_photo_url")
    private String checkInPhotoUrl;

    @Column(name = "check_out_photo_url")
    private String checkOutPhotoUrl;

    @Column(name = "check_in_location")
    private String checkInLocation; // GPS coordinates

    @Column(name = "check_out_location")
    private String checkOutLocation;

    @Column(length = 500)
    private String notes;

    @Column(name = "face_recognition_confidence")
    private Double faceRecognitionConfidence;

    @Column(name = "is_verified")
    private Boolean isVerified = true;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // Tính tổng số giờ làm việc
    public Double getWorkingHours() {
        if (checkInTime != null && checkOutTime != null) {
            long minutes = java.time.Duration.between(checkInTime, checkOutTime).toMinutes();
            return minutes / 60.0;
        }
        return 0.0;
    }
}

